cdc/generation.cc

```diff
@@ -328,4 +356,290 @@ void update_streams_description(
+        return;
+    }
+
+    cdc_log.warn("No CDC generation seen during startup. Upgrading?");
+
+    try {
+        new_generation({}, in_single_node_test);
+    } catch (...) {
+        cdc_log.error(
+            "Could not create new generation on startup. Reason: {}\n"
+            " If you are performing a rolling upgrade, make sure you restart nodes one by node,"
+            " and make sure that nodes are connected (there are no network issues)."
+            " You can still proceed with the upgrade. One of the nodes should eventually be able"
+            " to create a CDC generation. If every node reports this error, try restarting one of the nodes"
+            " so it can retry.",
+            std::current_exception());
+        // FIXME: after implementing check_and_repair_cdc_generation(), mention it here.
+    }
+}
+
+void generation_service::update_generation_timestamp(db_clock::time_point ts) {
+    assert_shard_zero(__PRETTY_FUNCTION__);
+
+    auto units = get_units(_current_ts_sem, 1).get0();
+
+    if (_current_ts && *_current_ts >= ts) {
+        return;
+    }
+
+    db::system_keyspace::update_cdc_streams_timestamp(ts).get();
+    _gossiper.add_local_application_state(
+            gms::application_state::CDC_STREAMS_TIMESTAMP, gms::versioned_value::cdc_streams_timestamp(ts)
+    ).get();
+    _current_ts = ts;
+
+    cdc_log.info("Updated latest generation timestamp to {}.", *_current_ts);
+}
+
+void generation_service::restore_generation_timestamp() {
+    assert_shard_zero(__PRETTY_FUNCTION__);
+    assert(db::system_keyspace::bootstrap_complete());
+
+    auto units = get_units(_current_ts_sem, 1).get0();
+
+    if (_current_ts) {
+        return;
+    }
+
+    cdc_log.info("Restoring generation timestamp from local tables...");
+
+    auto ts = db::system_keyspace::get_saved_cdc_streams_timestamp().get0();
+    if (!ts) {
+        cdc_log.info("No generation timestamp found in local tables.");
+        return;
+    }
+
+    _gossiper.add_local_application_state(
+            gms::application_state::CDC_STREAMS_TIMESTAMP, gms::versioned_value::cdc_streams_timestamp(*ts)
+    ).get();
+    _current_ts = ts;
+
+    cdc_log.info("Generation timestamp restored: {}", *_current_ts);
+}
+
+void generation_service::on_change(gms::inet_address ep, gms::application_state state, const gms::versioned_value& v) {
+    assert_shard_zero(__PRETTY_FUNCTION__);
+
+    if (state != gms::application_state::CDC_STREAMS_TIMESTAMP) {
+        return;
+    }
+
+    auto ts = gms::versioned_value::cdc_streams_timestamp_from_string(v.value);
+    cdc_log.debug("Endpoint: {}, CDC generation timestamp change: {}", ep, ts);
+    if (!ts) {
+        return;
+    }
+
+    learn_generation(*ts);
+}
+
+void generation_service::on_join(gms::inet_address ep, gms::endpoint_state ep_state) {
+    assert_shard_zero(__PRETTY_FUNCTION__);
+
+    for (const auto& e : ep_state.get_application_state_map()) {
```

Comment:
Do we have to loop over all the application states? Can't we just get CDC_STREAMS_TIMESTAMP?

Reply:
fixd